tests/gnrc_dhcpv6_client/tests-as-root/01-run.py

```diff
@@ -8,24 +8,73 @@
 
 import sys
 from testrunner import run
+from ipaddress import (
+    IPv6Address,
+    IPv6Network,
+)
+
+
+IA_NA_ADDRESS_POOL_PREFIX = "2001:db8:1::"
 
 
 def testfunc(child):
     child.expect(r"Iface\s+\d+")
     child.expect(r"inet6 addr:\sfe80:[0-9a-f:]+\s+scope: link")
     child.expect(r"Iface\s+\d+")
     child.expect(r"inet6 addr:\s+fe80:[0-9a-f:]+\s+scope: link")
-    child.expect(r"inet6 addr:\s+(?P<global_addr>[0-9a-f:]+)\s+scope: global")
-    global_addr = child.match.group("global_addr")
+
+    global_addr_1, global_addr_2 = extract_global_addresses(child)
+
+    global_pfx = extract_global_prefix(child)
+
+    test_global_addrs(global_addr_1, global_addr_2, global_pfx)
+
+
+def extract_global_prefix(child):
     child.expect(r"(?P<global_pfx>[0-9a-f:]+)/64\s+dev #\d\s+"
                  r"expires \d+ sec\s+"
                  r"deprecates \d+ sec")
     global_pfx = child.match.group("global_pfx")
+
     if global_pfx.endswith("::"):
         # remove one trailing : in case there are no 0s between prefix and
         # suffix
         global_pfx = global_pfx[0:-1]
-    assert global_addr.startswith(global_pfx)
+
+    return global_pfx
+
+
+def extract_global_address(child):
+    """Expect and extract a global address from the command line."""
+    child.expect(r"inet6 addr:\s+(?P<global_addr>[0-9a-f:]+)\s+scope: global")
+    return child.match.group("global_addr")
+
+
+def extract_global_addresses(child):
+    """Extract two global addresses and return them as a tuple."""
+    return extract_global_address(child), extract_global_address(child)
+
+
+def check_ia_na_addr(ia_na_addr):
+    """Check if the expected IA_NA address has been assigned"""
+    return IPv6Address(ia_na_addr) in IPv6Network("{}/64".format(IA_NA_ADDRESS_POOL_PREFIX))
+
+
+def check_ia_pd_addr(ia_pd_addr, global_pfx):
+    """Check if the expected IA_PD address has been assigned"""
+    return ia_pd_addr.startswith(global_pfx)
+
+
+def check_global_addrs(ia_na_addr, ia_pd_addr, global_pfx):
+    """Perform IA_NA check for the first and IA_PD for the second address"""
+    return check_ia_na_addr(ia_na_addr) and check_ia_pd_addr(ia_pd_addr, global_pfx)
+
+
+def test_global_addrs(global_addr_1, global_addr_2, global_pfx):
+    """Assert that one global address is the IA_NA and the other one is the IA_PD address"""
+    result_1 = check_global_addrs(global_addr_1, global_addr_2, global_pfx)
+    result_2 = check_global_addrs(global_addr_2, global_addr_1, global_pfx)
+    assert result_1 != result_2
```

Comment:
And this way, this check is still valid.

Reply:
Regarding the [comment above](https://github.com/RIOT-OS/RIOT/pull/16228/files#r686297272), do you mean adding an additional assertion

```suggestion
    assert result_1 != result_2
    assert any(result_1.values()) != any(result_2.values())
```

or replacing the existing one?

```suggestion
    assert any(result_1.values()) != any(result_2.values())
```

Reply:
Both `any`s should be true (as both value lists should contain at least one true value), so comparing them does not make sense.

```suggestion
    assert any(result_1.values())
    assert any(result_2.values())
    assert result_1 != result_2
```

Reply:
Maybe also throw in some comments to explain the crazy predicate logic we are doing there ;-).

Reply:
> as both value lists should contain at least one true value

Ah, I see! I am actually not sure if this is the case as `check_global_addrs` should only return `True` (or two `True` values) if the first address has been assigned by IA_NA and the second one by IA_PD (or – in the case of `result_2` – vice versa as the inputs are flipped). Otherwise you should get a `False` (or two `False` values). Therefore using

```python
assert any(result_1.values())
assert any(result_2.values())
```

should actually result in a failed assertion in one (and only one) of the two cases. Therefore, I think using only 

```python
assert result_1 != result_2
```

should be sufficient if the rest of the test implemetation is correct, with the extended return value of `check_global_addrs` providing addtional safety. I hope I am not missing anything, though :D

Reply:
In that case the test seems to be flawed.. One check should check, if the global address was set by IA_NA and one by IA_PD. As such only one of them should be true in either call of `check_global_addrs`, and the results should be disjunct.

Reply:
> with the extended return value of check_global_addrs providing addtional safety

Hmm, now that I think about it, the lazy evaluation might actually be more accurate to ensure correct behavior.

Consider that the first address is the correct IA_NA address and the other one is not the expected IA_PD address. In the case where only a single boolean value is returned, both `check_global_addrs` calls return `False` which leads to a failed assertion.

In the case with the dictionary (and I guess similarly in the case of the list) this would result in 

```python
assert {"ia_na_check": True, "ia_pd_check": False} != {"ia_na_check": False, "ia_pd_check": False}
```

which should pass. I guess I would therefore either stick to the original solution or add

```python
assert any(result_1.values()) != any(result_2.values())
```
as mentioned above.

Reply:
> In that case the test seems to be flawed.. One check should check, if the global address was set by IA_NA and one by IA_PD. As such only one of them should be true in either call of `check_global_addrs`, and the results should be disjunct.

Oh, I see, we might have two different approaches to this. Currently, I am checking if either 

* the first address has been assigned by IA_NA and the second one by IA_PD or
* the first address has been assigned by IA_PD and the second one by IA_NA.

The assertion checks that exactly one of these two conditions is fulfilled while the other one is false. Otherwise it fails.

Reply:
> Oh, I see, we might have two different approaches to this. Currently, I am checking if either
> 
>     * the first address has been assigned by IA_NA and the second one by IA_PD or
> 
>     * the first address has been assigned by IA_PD and the second one by IA_NA.

Then I don't understand this:



> In the case with the dictionary (and I guess similarly in the case of the list) this would result in
> 
> ```python
> assert {"ia_na_check": True, "ia_pd_check": False} != {"ia_na_check": False, "ia_pd_check": False}
> ```

Shouldn't the second dictionary be `{"ia_na_check": False, "ia_pd_check": True}` in that case?

Reply:
Ahh now I get what you are trying to do

From my testing with

<details><summary>this patch applied</summary>

```diff
diff --git a/tests/gnrc_dhcpv6_client/kea-dhcp6.conf b/tests/gnrc_dhcpv6_client/kea-dhcp6.conf
index 593edcb9d8..a2a8ab429c 100644
--- a/tests/gnrc_dhcpv6_client/kea-dhcp6.conf
+++ b/tests/gnrc_dhcpv6_client/kea-dhcp6.conf
@@ -23,24 +23,9 @@
     {    "interface": "tapbr0",
          "subnet": "2001:db8::/32",
          "pools": [ { "pool": "2001:db8:1::/64" } ],
-         "pd-pools": [ { "prefix": "2001:db8::",
-                         "prefix-len": 32,
+         "pd-pools": [ { "prefix": "2001:db8:8000::",
+                         "prefix-len": 33,
                          "delegated-len": 64 } ] }
     ]
-  },
-"Logging":
-{
-  "loggers": [
-    {
-      "name": "kea-dhcp6",
-      "output_options": [
-          {
-            "output": "stderr"
-          }
-      ],
-      "severity": "INFO",
-      "debuglevel": 0
-    }
-  ]
-}
+  }
 }
diff --git a/tests/gnrc_dhcpv6_client/tests-as-root/01-run.py b/tests/gnrc_dhcpv6_client/tests-as-root/01-run.py
index 8dec38935c..4177708463 100755
--- a/tests/gnrc_dhcpv6_client/tests-as-root/01-run.py
+++ b/tests/gnrc_dhcpv6_client/tests-as-root/01-run.py
@@ -57,12 +57,16 @@ def extract_global_addresses(child):
 
 def check_ia_na_addr(ia_na_addr):
     """Check if the expected IA_NA address has been assigned"""
-    return IPv6Address(ia_na_addr) in IPv6Network("{}/64".format(IA_NA_ADDRESS_POOL_PREFIX))
+    res = IPv6Address(ia_na_addr) in IPv6Network("{}/64".format(IA_NA_ADDRESS_POOL_PREFIX))
+    print("check_ia_na_addr", ia_na_addr, res)
+    return res
 
 
 def check_ia_pd_addr(ia_pd_addr, global_pfx):
     """Check if the expected IA_PD address has been assigned"""
-    return ia_pd_addr.startswith(global_pfx)
+    res = ia_pd_addr.startswith(global_pfx)
+    print("check_ia_pd_addr", ia_pd_addr, global_pfx, res)
+    return res
 
 
 def check_global_addrs(ia_na_addr, ia_pd_addr, global_pfx):
@@ -72,7 +76,9 @@ def check_global_addrs(ia_na_addr, ia_pd_addr, global_pfx):
 
 def test_global_addrs(global_addr_1, global_addr_2, global_pfx):
     """Assert that one global address is the IA_NA and the other one is the IA_PD address"""
+    print("result_1")
     result_1 = check_global_addrs(global_addr_1, global_addr_2, global_pfx)
+    print("result_2")
     result_2 = check_global_addrs(global_addr_2, global_addr_1, global_pfx)
     assert result_1 != result_2
```

(ignore the logging part removal that is because my kea is quite new and not understood by that version anymore)

</details>

I see, that the results could be

```
assert {"ia_na_check": True, "ia_pd_check": True} != {"ia_na_check": False, "ia_pd_check": False}
```

(which makes more sense than `ia_pd_check` always being `False`), so the final check actually should be

```py
assert result_1["ia_na_check"] or result_2["ia_na_check"]
assert result_1["ia_pd_check"] or result_2["ia_pd_check"]
```

I really recommend changing the prefix for PD (and maybe adopting the check to be similar to the PD one, as the IA_NA's prefix also ends up in the prefix list.